Enhanced Debug.js

bbert opened the pull request:
Enhanced Debug.js that adds possibility to insert callee object name and media type in the output message, which is very useful.
This would facilitate debugging by enabling filtering log messages.
An example is provided within BufferController.js, but if the PR is merged we could adapt all object outputting log messages.

KozhinM commented:
I like the idea of adding callee name to the log, but I think that approach should be more generic. I mean, instead of manually  putting getName/getType all over the code such methods have to be somehow inherited/extended by the class. This would require more  work and probably revising the current approach for creating object. Probably a kind of a factory/wrapper for dijon lib. For example, something like:
```
//ObjectFactory.js
ObjectFactory.getObject = function(className) {
	var obj = this.system.getObject(className);
	//extend the object with dash.js specific methods
	obj.getName = function() {return  className};
	return obj;
}
```

Also it would be nice to have an option to show/hide callee  for example by Debug.showCalleeName(true), i.e. Debug class has to be responsible for displaying that information, not the callee object itself.
 
Finally, checking for getType in Debug.log is not reliable because there is no strict rules declaring that getType function must return media type (audio/video/text).

bbert commented:
Yes I agree that a more generic approach is much better.
For the getName, we can also redefine the getObject function, something like:

    var _getObject = dijon.System.prototype.getObject;
    dijon.System.prototype.getObject = function (name) {
        var obj = _getObject.call(this, name);
        if (typeof obj === "object" && !obj.getName) {
            obj.getName = function () {return name;};
        }
        return obj;
    };

I tried and it works.

OK also for adding Debug.showCalleeName function.

For the getType (or maybe getMediaType is better), I do not have any other solution. But this would be very helpful having automatically the media type in the messages instead of adding manually the type in the log messages as it is currently done. Then we could have for example:

    [bufferController][video] Buffered range: 0 - 12.011999
    [bufferController][audio] Buffered range: 0.005458 - 24.026791

thus, filtering/isolating log messages according to media type would facilitate debugging.

KozhinM commented:
The idea with overriding getObject looks fine, still I have a feeling that we may think about removing explicit use of dijon's system.getObject and replace it with a kind of a wrapper for object creation logic.Of course, this is out of scope of this issue and for now your approach may serve.

Regarding getMediaType - I agree that it is helpful to add a type in a log message. My point was to replace a quite generic "getType" with more specific "getMediaType". Also we may change getObject so that it would add setMediaType/getMediaType methods just like getName. Once an object is created we will call setMediaType for these objects that support it. As a default value we could use "stream" which indicates that the class is common for both audio/video (like singletons).

AkamaiDASH commented:
Looks good and seems very helpful.  I agree with @KozhinM  last comments. I really think it needs to have the api to toggle on/off  - Also Just want to make sure the comma separated log items is maintained and it looks like it is.

bbert commented:
Agree with your comments. But for the time being I suggest to keep dijon as it is and simply overload system.getObject function.

I pushed some new code:
- Added (in dijon's system.getObject overloaded function) set/getMediaType() function (see MediaPlayer.js)
- Added showCalleeName() function to Debug.js
- Modified debug.log calls to add callee object as a parameter.

KozhinM commented:
Great! Much better. Still I have a few notes to make it perfect :)
	1. Some logs have already had "type" in message itself. I think it has to be removed to avoid duplicates, e.g.:
	`[scheduleController][video] Stalling video Buffer`
	2. It would be nice if we did not need to put this/self as a first argument of log function each time it is called. I can think only about one relatively decent solution for this:
	In Debug.js add the following:
```
setup: function() {
    this.system.mapValue('log', this.log);
}
```
Update Debug.log:
```
log: function() {
    …
    if (showCalleeName) {
	message += "[" + this.getName() + "]"
    }
    ...
}
```

Now if you add `log: undefined` outlet to any class you can use it w/o accessing Debug object, i.e. `this.log("some useful info");` instead of `this.debug.log("some useful info");`.  'this' in log function will point to the callee object instead of Debug object. Thus, there is no need to pass this/self to log.

@AkamaiDASH, what do you think?

bbert commented:
Great and thanks!
I was not yet familiar enough with dijon to think about this solution. This is much better this way.
I have pushed code with this solution.
Also, in order to avoid duplicating the "type" in the messages, I decided to insert it anyway in Debug.js's log function, even if showCalleeName is set to false, e.g.:

    [video] Stalling Buffer

KozhinM commented:
Looks good, thanks! Just one last thing - I see that some of jasmine tests fail. In dev branch all of them pass. Could you please take a look at them? I am going to merge this PR as soon as all the tests pass.

bbert commented:
I corrected so that all the tests pass. Please check.

KozhinM commented:
Great! I think it is time to merge this PR. @bbert, thanks for helpful update!

AkamaiDASH commented:
@greg80303 would you like to merge or shall I ?
